### ISSUE TITLE
clang crash on linear code 

### ISSUE THREAD

**Issue by FROL256** [vgvassilev/clad#575]
Use latest release 1.1

code:

```
static inline void mat4_colmajor_mul_vec4(float* __restrict RES, const float* __restrict B, const float* __restrict V) // modern gcc compiler succesfuly vectorize such implementation!
{
	RES[0] = V[0] * B[0] + V[1] * B[4] + V[2] * B[ 8] + V[3] * B[12];
	RES[1] = V[0] * B[1] + V[1] * B[5] + V[2] * B[ 9] + V[3] * B[13];
	RES[2] = V[0] * B[2] + V[1] * B[6] + V[2] * B[10] + V[3] * B[14];
	RES[3] = V[0] * B[3] + V[1] * B[7] + V[2] * B[11] + V[3] * B[15];
}

static inline void mul_mat_mat(const float m1[16], const float m2[16], float mRes[16])
{
  mat4_colmajor_mul_vec4(mRes+0 , m1,  m2+0 );
  mat4_colmajor_mul_vec4(mRes+4 , m1,  m2+4 );
  mat4_colmajor_mul_vec4(mRes+8 , m1,  m2+8 );
  mat4_colmajor_mul_vec4(mRes+12, m1,  m2+12);
}

static constexpr int POS_X = 0;
static constexpr int POS_Y = 1;
static constexpr int POS_Z = 2;

static constexpr int ROT_X = 3;
static constexpr int ROT_Y = 4;
static constexpr int ROT_Z = 5;

float transformX(const float a_data[6], const float vInput[4]) 
{
  const float cosX = std::cos(a_data[ROT_X]);
  const float sinX = std::sin(a_data[ROT_X]); 
  const float cosY = std::cos(a_data[ROT_Y]);
  const float sinY = std::sin(a_data[ROT_Y]); 
  const float cosZ = std::cos(a_data[ROT_Z]);
  const float sinZ = std::sin(a_data[ROT_Z]); 

  float mtran[16] = {1,0,0,0,
                     0,1,0,0,
                     0,0,1,0,
                     a_data[POS_X], a_data[POS_Y], a_data[POS_Z],1};
  float mRotX[16] = {1,0,0,0,
                     0,cosX,sinX,0,
                     0,-sinX,cosX,0,
                     0,0,0,1};
  float mRotY[16] = {cosY,0,-sinY,0,
                     0,1,0,0,
                     sinY,0,cosY,0,
                     0,0,0,1}; 
  float mRotZ[16] = {cosZ,sinZ,0,0,
                     -sinZ,cosZ,0,0,
                     0,0,1,0,
                     0,0,0,1};                                             
  
  float mRes[16];
  mul_mat_mat(mRotZ, mRotY, mRes);  // mRes  := mRotZ*mRotY
  mul_mat_mat(mRotX, mRes,  mRotZ); // mRotZ := mRotX*(mRotZ*mRotY)
  mul_mat_mat(mtran, mRotZ, mRes);  // mRes  := mtran*(mRotX*(mRotZ*mRotY))
  
  float res[4];
  mat4_colmajor_mul_vec4(res, mRes, vInput);

  return res[0];  
} 
```

clang-14 -I /home/frol/PROG/clad_autodiff_examples/inst/include -x c++ -std=c++11 -g -fplugin=/home/frol/PROG/clad_autodiff_examples/inst/lib/clad.so SourceFile.cpp -lstdc++ -o z_out -lstdc++ -lm -Xclang -plugin-arg-clad -Xclang -fgenerate-source-file -DCLAD_NO_NUM_DIFF 

The whole bug-report data is located here:
https://drive.google.com/file/d/10xDbs9V-2jAbW5o6tegzBElDZb_Z0PLC/view?usp=drive_link

**Comment by parth-07** [comment]
Hi @FROL256,
Can you please give me access to the google drive link you shared in the issue description?

**Comment by FROL256** [comment]
I have opened access, srry

**Comment by parth-07** [comment]
@FROL256 Thank you, I will check the issue soon.

**Comment by vgvassilev** [comment]
@parth-07 ping.

**Comment by parth-07** [comment]
Hi @FROL256,

Thank you for creating the issue and providing all the details.

I am sorry for the late reply.

Two issues in Clad need to be fixed to support the code you are trying to differentiate.

1) Supporting `const T` types in `clad::array<T>`. Currently, `clad::array<const float>` is being instantiated at a few places when differentiating your code. But `clad::array<T>` currently doesn't correctly handle `const` types. This can be fixed relatively easily.

2) Supporting pointer arithmetic. Currently, your code uses `mRes+4`, `m2+4` and so on. Reverse-mode does not have pointer arithmetic support as of now.

**Comment by FROL256** [comment]
Hy @parth-07. Thank you very much for your reply. I'll try to rewrite the code.

**Comment by parth-07** [comment]
Fixed by #763.
